### PR TITLE
projecting nft metadata

### PR DESCRIPTION
**Pull request by greatertomi**
# Context

Personalized handles have their metadata defined in datums (cip68/222), so we’ll need to update our AssetProvider to resolve metadata from this new source. In order to do that a new projection is needed

# Proposed Solution

This adds a new `withNftMetadata` mapper to project NFT metadata and a `withCIP67` mapper to properly format CIP67 assets. `storeNftMetadata` operator was added to `projection-typeorm` to store NFT metadata in postgres DB

# Important Changes Introduced

### CODE REVIEW

File: packages/projection/src/operators/Mappers/withMint.ts [hunk 1]
```diff
@@ -21,13 +22,14 @@ export interface WithMint {
...
   ...evt,
-  mint: evt.block.body.flatMap(({ body: { mint }, auxiliaryData }, txIndex) =>
+  mint: evt.block.body.flatMap(({ body: { mint, outputs }, auxiliaryData }, txIndex) =>
     [...(mint?.entries() || [])].map(
       ([assetId, quantity]): Mint => ({
         assetId,
         assetName: Asset.AssetNameLabel.decode(Cardano.AssetId.getAssetName(assetId)),
         assetNameHex: Cardano.AssetId.getAssetName(assetId),
         compactTxId: computeCompactTxId(evt.block.header.blockNo, txIndex),
+        outputs,
```
Comment (mkazlauskas) [inline packages/projection/src/operators/Mappers/withMint.ts:32]: I don't think the asset has to be **minted**. You could probably recreate the datum by just transferring the tokens back to the script:
```Similarly, if the issuer wants the NFTs/FTs to evolve or wants a mechanism to update the metadata, the reference NFT can be locked at the address of a script with arbitrary logic that the issuer decides.```

It means that we need to check **all** outputs in every transaction, not just in the transactions that some assets were **minted**

File: packages/projection/src/operators/Mappers/withNftMetadata.ts [hunk 2]
```diff
@@ -11,45 +11,36 @@ export interface WithNftMetadata {
...
+    evt.utxo.produced.flatMap(([, txOutUtxo]): Array<readonly [Cardano.AssetId, Asset.NftMetadata]> | null => {
+      const assets = txOutUtxo.value.assets;
+      if (!assets || assets.size === 0) return null;
+      const nftMetadata = Asset.NftMetadata.fromPlutusData(txOutUtxo.datum, dummyLogger);
+      if (!nftMetadata) return null;
+      return [...assets]
         .filter(
-          ({ assetName }) =>
-            !assetName || assetName?.label === Asset.AssetNameLabelNum.UserNFT || assetName?.label === KORALAB_NFT
+          ([, { assetName }]) => assetName.label === Asset.AssetNameLabelNum.UserNFT || assetName?.label === KORALAB_NFT
```
Comment (mkazlauskas) [inline packages/projection/src/operators/Mappers/withNftMetadata.ts:23]: I think the datum is stored together with the ReferenceNFT (not UserNFT). Please double-check in cip68

Comment (mkazlauskas) [reply packages/projection/src/operators/Mappers/withNftMetadata.ts:23]: Also we shouldn't hardcode KORALAB_NFT, but rather inject the labels via an argument (see example of such operator in filterProducedUtxoByAssetPolicyId)

File: packages/projection/src/operators/Mappers/withUtxo.ts [hunk 3]
```diff
@@ -24,6 +24,19 @@ export interface WithUtxo {
   };
 }
 
+const mapAssetValueToDecodedAssetName = (value: Cardano.Value) => ({
+  ...value,
+  assets: new Map(
+    [...(value?.assets || [])].map(([assetId, quantity]) => {
+      const assetValue = {
+        assetName: Asset.AssetNameLabel.decode(Cardano.AssetId.getAssetName(assetId)),
```
Comment (mkazlauskas) [inline packages/projection/src/operators/Mappers/withUtxo.ts:32]: let's include the encoded asset name here as well

File: packages/core/src/Asset/cip67.ts [hunk 4]
```diff
@@ -15,6 +15,13 @@ export enum AssetNameLabelNum {
 export type AssetNameLabel = OpaqueNumber<'AssetNameLabelNum'>;
 export const AssetNameLabel = (value: number): AssetNameLabel => value as AssetNameLabel;
 
+export const AssetNameLabelObject = {
```
Comment (mkazlauskas) [inline packages/core/src/Asset/cip67.ts:18]: Let's delete the enum and keep only this object, named AssetNameLabelNum

File: packages/projection/src/operators/Mappers/withNftMetadata.ts [hunk 5]
```diff
@@ -1,37 +1,31 @@
+/* eslint-disable import/no-extraneous-dependencies */
 import { Asset, Cardano } from '@cardano-sdk/core';
 import { WithCIP67 } from './withCIP67';
 import { WithMint } from './withMint';
 import { WithUtxo } from './withUtxo';
-import { dummyLogger } from 'ts-log';
 import { isNotNil } from '@cardano-sdk/util';
+import { logger } from '@cardano-sdk/util-dev';
```
Comment (mkazlauskas) [inline packages/projection/src/operators/Mappers/withNftMetadata.ts:7]: `util-dev` package should only be used for tests. Please inject a `Logger` in the argument to `withNftMetadata`

File: packages/core/src/Asset/cip67.ts (hunk 4 above)
Comment (mkazlauskas) [reply packages/core/src/Asset/cip67.ts]: a6a8d94571120839f69de94acd7f3985e5be99e4

File: packages/projection-typeorm/test/operators/storeNftMetadata.test.ts [hunk 6]
```diff
@@ -0,0 +1,427 @@
...
+  });
+
+  it('can store and retrieve files', async () => {
+    const events = await lastValueFrom(project$(withInlineDatumEvents).pipe(toArray()));
+    expect(events.length).toBeGreaterThan(1);
+    expect(await nftMetadataRepo.count()).toBeGreaterThan(0);
+    const nftAssets = await nftMetadataRepo.find({ select: ['files'] });
+    expect(nftAssets.length).toBeGreaterThan(0);
+    expect(nftAssets[0].files).toMatchObject([
+      { mediaType: 'image/gif', name: 'Shrine #2142', src: 'ipfs://QmVTzRmbmxQS3NkPcqkxjSLR4Q2oBdMbF1qtAqhJXET752' },
```
Comment (mkazlauskas) [inline packages/projection-typeorm/test/operators/storeNftMetadata.test.ts:422]: Let's make this less brittle, e.g.
```suggestion
    const file = nftAssets.find(asset => asset.files?.length > 0)!.files![0];
    expect(typeof file.mediaType).toBe('string')
    ...
```

File: packages/projection/test/operators/Mappers/withNftMetadata.test.ts [hunk 7]
```diff
@@ -0,0 +1,291 @@
...
+            auxiliaryData: undefined
+          }))
+        }
+      }))
+    );
+
+const isDatumInBlock = (data: Cardano.Block) => {
+  if (!data || !data.body || !Array.isArray(data.body)) {
+    return false; // Invalid data format
+  }
```
Comment (mkazlauskas) [inline packages/projection/test/operators/Mappers/withNftMetadata.test.ts:29]: Is there any reason to check this? The type specifies that those properties are required
```suggestion
```

**Review by mkazlauskas — CHANGES_REQUESTED** [review]

File: packages/projection/test/operators/Mappers/withNftMetadata.test.ts [hunk 8]
```diff
@@ -0,0 +1,320 @@
...
+      version: '1.0'
+    };
+
+    expect(nftMetadata).toMatchObject([
+      { nftMetadata: result1, userTokenAssetId: assetId1 },
+      { nftMetadata: result2, userTokenAssetId: assetId2 }
+    ]);
+  });
+
+  it('to maps all minted nfts, and handles cip 67 asset names correctly', async () => {
```
Comment (mkazlauskas) [inline packages/projection/test/operators/Mappers/withNftMetadata.test.ts:198]:
```suggestion
  it('finds cip68 nft metadata', async () => {
```

File: packages/projection/test/operators/Mappers/withNftMetadata.test.ts [hunk 9]
```diff
@@ -0,0 +1,320 @@
...
+          ['characters', 'letters,numbers'],
+          ['numeric_modifiers', ''],
+          ['version', 1n]
+        ]),
+        version: '1'
+      }
+    ]
+  ]);
+
+  it('maps all minted nft into a map', async () => {
```
Comment (mkazlauskas) [inline packages/projection/test/operators/Mappers/withNftMetadata.test.ts:163]:
```suggestion
  it('finds cip25 nft metadata', async () => {
```

File: packages/projection/test/operators/Mappers/withNftMetadata.test.ts [hunk 10]
```diff
@@ -0,0 +1,320 @@
...
+        Mappers.withCIP67(),
+        Mappers.withNftMetadata({ logger: dummyLogger })
+      )
+    );
+
+    expect(nftMetadata).toMatchObject(someNftMetadata);
+    expect(nftMetadata).toMatchObject([datumNftmetadata]);
+  });
+
+  it('should handle non minted cip67 assets properly', async () => {
```
Comment (mkazlauskas) [inline packages/projection/test/operators/Mappers/withNftMetadata.test.ts:220]:
```suggestion
  it('finds updated cip68 nft metadata (asset was not minted in this block)', async () => {
```

**Review by mkazlauskas — CHANGES_REQUESTED** [review]

File: packages/projection/test/operators/Mappers/withNftMetadata.test.ts (hunk 7 above)
Comment (greatertomi) [reply packages/projection/test/operators/Mappers/withNftMetadata.test.ts]: This checks that the block has the proper structure before checking a datum. Removed on b37531e

File: packages/projection/test/operators/Mappers/withNftMetadata.test.ts (hunk 8 above)
Comment (greatertomi) [reply packages/projection/test/operators/Mappers/withNftMetadata.test.ts]: Fixed on b37531e

File: packages/projection/test/operators/Mappers/withNftMetadata.test.ts (hunk 9 above)
Comment (greatertomi) [reply packages/projection/test/operators/Mappers/withNftMetadata.test.ts]: Fixed on b37531e

File: packages/projection/test/operators/Mappers/withNftMetadata.test.ts (hunk 10 above)
Comment (greatertomi) [reply packages/projection/test/operators/Mappers/withNftMetadata.test.ts]: Fixed on b37531e

File: packages/e2e/test/projection/single-tenant-utxo.test.ts [hunk 11]
```diff
@@ -47,7 +47,8 @@ const createDataSource = () =>
       Postgres.BlockDataEntity,
       Postgres.AssetEntity,
       Postgres.TokensEntity,
-      Postgres.OutputEntity
+      Postgres.OutputEntity,
+      Postgres.NftMetadataEntity
```
Comment (mkazlauskas) [inline packages/e2e/test/projection/single-tenant-utxo.test.ts:51]: This is the cost of having `AssetEntity.nftMetadata` column: all relations require you to explicitly specify the entity. Perhaps there is some workaround we could implement, but I think for now it's fine.

File: packages/projection-typeorm/src/createDataSource.ts [hunk 12]
```diff
@@ -71,6 +71,9 @@ const namingOverrides: Partial<NamingStrategyInterface> = {
   },
   tableName(targetName, userSpecifiedName) {
     return userSpecifiedName || toTableName(snakeCase(targetName));
+  },
+  uniqueConstraintName(tableName, columnNames) {
+    return `UQ_${toTableName(tableName)}_${columnNames.join('_')}}`;
```
Comment (iccicci) [inline packages/projection-typeorm/src/createDataSource.ts:76]: Could `UC` be the right prefix for unique constraints?

Comment (mkazlauskas) [reply packages/projection-typeorm/src/createDataSource.ts:76]: Sure if it's the more common convention. TypeORM used UQ by default so I left that, just replaced the hex hash with readable names

Comment (iccicci) [reply packages/projection-typeorm/src/createDataSource.ts:76]: Nice to learn about the TypeORM default. Thank you

File: packages/e2e/test/projection/single-tenant-utxo.test.ts (hunk 11 above)
Comment (rhyslbw) [reply packages/e2e/test/projection/single-tenant-utxo.test.ts:51]: Can we resolve this conversation @mkazlauskas @greatertomi ?

File: packages/cardano-services/src/Handle/TypeOrmHandleProvider.ts [hunk 13]
```diff
@@ -43,6 +43,10 @@ export class TypeOrmHandleProvider extends TypeormProvider implements HandleProv
 
       const mapEntity = (entity: PartialHandleEntity | undefined): HandleResolution | null => {
         if (!entity) return null;
+        if (!entity.cardanoAddress) {
+          this.logger.warn(`Invalid handle '${entity.handle}'. Double mint?`);
```
Comment (rhyslbw) [inline packages/cardano-services/src/Handle/TypeOrmHandleProvider.ts:47]:
```suggestion
          this.logger.warn(`${entity.handle} has no associated address, which could be the result of a double mint.`);
```

File: packages/projection-typeorm/src/operators/storeNftMetadata.ts [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+    block: {
+      header: { slot }
+    },
+    mint
+  }: StoreNftMetadataEvent,
+  { nftMetadataRepository, assetRepository }: HandlerDependencies
+) => {
+  const nftMetadataEntities = await Promise.all(nftMetadata.map(toNftMetadataEntity(slot, assetRepository)));
+  const storedNftMetadata = await nftMetadataRepository.insert(nftMetadataEntities);
+  // Perf: see if it's possible to build a single query that updates all assets
```
Comment (rhyslbw) [inline packages/projection-typeorm/src/operators/storeNftMetadata.ts:56]: Is this captured in our backlog?

**Review by rhyslbw — CHANGES_REQUESTED** [review]

File: packages/core/test/Serialization/PlutusData.test.ts [hunk 15]
```diff
@@ -8,4 +9,179 @@ describe('PlutusData', () => {
     const fromCbor = PlutusData.fromCbor(cbor);
     expect(fromCbor.toCore()).toEqual(plutusData);
   });
+
+  it('converts (TODO: describe is special about this that fails) inline datum', () => {
```
[the remark this answers is not in the document]
Comment (rhyslbw) [reply packages/core/test/Serialization/PlutusData.test.ts]: https://github.com/input-output-hk/cardano-js-sdk/pull/814/commits/389c5e2a75589f536ca8fcf764b45e2c11524532

[the remark this answers is not in the document]
Comment (mkazlauskas) [reply packages/core/test/Serialization/PlutusData.test.ts]: Renamed the test in 389c5e2a75589f536ca8fcf764b45e2c11524532

File: packages/projection-typeorm/src/operators/storeNftMetadata.ts (hunk 14 above)
Comment (mkazlauskas) [reply packages/projection-typeorm/src/operators/storeNftMetadata.ts:56]: Created LW-7705

File: packages/cardano-services/src/Handle/TypeOrmHandleProvider.ts (hunk 13 above)
Comment (mkazlauskas) [reply packages/cardano-services/src/Handle/TypeOrmHandleProvider.ts]: 53e0809bb51e3a94d9ad4a396e0e4d828ae5e077

**Review by rhyslbw — APPROVED** [review]
Amazing work here guys!  It's great to see the new serialisation functionality integrated, to not only solve our immediate blocker, but also to bring more value for other projects.

**Pull request merged**